crates/bifrost/src/types.rs

```diff
@@ -15,29 +15,45 @@ use crate::loglet::LogletOffset;
 use restate_types::logs::{Lsn, SequenceNumber};
 use serde::{Deserialize, Serialize};
 
+// Only implemented for LSNs
 pub(crate) trait LsnExt: SequenceNumber {
     /// Converts a loglet offset into the virtual address (LSN).
+    #[track_caller]
     fn offset_by<S: SequenceNumber>(self, offset: S) -> Self {
         // This assumes that this will not overflow. That's not guaranteed to always be the
         // case but it should be extremely rare that it'd be okay to just wrap in this case.
         //
-        // We subtract from OLDEST because loglets start their offset from 1, not 0.
+        // We subtract from OLDEST because loglets might start offsets from a non-zero value.
         // 1 is the oldest valid offset within a loglet, 0 is an invalid offset.
         debug_assert!(offset != S::INVALID);
         let self_raw: u64 = self.into();
         let offset_raw: u64 = offset.into();
 
-        Self::from(self_raw.wrapping_add(offset_raw) - S::OLDEST.into())
+        Self::from(std::cmp::max(
+            Self::OLDEST.into(),
+            self_raw
+                .wrapping_add(offset_raw)
+                .saturating_sub(S::OLDEST.into()),
+        ))
     }
 
     /// Convert an LSN back to a loglet offset given a base_lsn.
+    #[track_caller]
     fn into_offset(self, base_lsn: Lsn) -> LogletOffset {
         let base_lsn_raw: u64 = base_lsn.into();
         let self_raw: u64 = self.into();
         let oldest_offset: u64 = LogletOffset::OLDEST.into();
+
+        if self_raw < base_lsn_raw {
+            return LogletOffset::OLDEST;
+        }
         // We must first add the oldest_offset before subtracting base_lsn_raw because self_raw
         // can be 0.
-        LogletOffset((self_raw + oldest_offset).saturating_sub(base_lsn_raw))
+        LogletOffset(
+            self_raw
+                .saturating_sub(base_lsn_raw)
+                .saturating_add(oldest_offset),
```

Comment:
Same question here regarding rather panicking instead of swallowing the problem?